src/provider/provider_coarse.c

```diff
@@ -606,11 +902,17 @@ static enum umf_result_t coarse_memory_provider_free(void *provider, void *ptr,
 
     // Merge with prev and/or next block if they are unused and have continuous
     // data.
-    block = block_merge_with_prev(block);
-    block_merge_with_next(block, &coarse_provider->block_list);
+    block = block_merge_with_prev(coarse_provider->free_blocks, block);
+    block = block_merge_with_next(coarse_provider->free_blocks, block,
+                                  &coarse_provider->block_list);
 
     assert(debug_check(coarse_provider));
 
+    int rv = free_blocks_add(coarse_provider->free_blocks, block);
```

Comment:
if the merging didn't happen, this func shouldn't be called, right? Move this call inside block_merge functions

Reply:
All is OK here. We are adding the "block", merged or not.